### ISSUE TITLE
unchanged minimap hud flags are sent to the client repeatedly

### ISSUE THREAD

**Issue by corarona** [minetest/minetest_game#2900]
The minimap hud flag is sent to the client every 5.3 seconds even if it has not changed in 

mods/map/init.lua:22

This issue is analogous to https://git.minetest.land/Mineclonia/Mineclonia/pulls/146 even though it is VERY limited here compared to mcl2/mcla so probably a minor issue. But I think it should check if the value is changed using player:hud_get_flags() before sending it to the client.

### Reproducing
- run minetest --info 2>&1 |grep TOCLIENT_HUD_SET_FLAGS
- enter a world with minetest_game
- do nothing for a few minutes
- observe in your terminal that hud flags are sent repeatedly even though nothing has changed 
- you can verify by commenting out lines 22-25 in the mentioned file that this is indeed caused by that

**Comment by appgurueu** [comment]
> > > But I think it should check if the value is changed using player:hud_get_flags()
> > 
> > 
> > I don't think it should. That should be done by the engine as `set_hud_flags()` is called. Mod code shouldn't be polluted.
> 
> ideally it would be checked by the engine, yes but people dont seem to want that !? [minetest/minetest#11459](https://github.com/minetest/minetest/issues/11459)

Then those same people probably won't want that in MTG either.

**Comment by erlehmann** [comment]
> Then those same people probably won't want that in MTG either.

I understand that there exist people who argue that the engine should send useless traffic all the time.

However, what I have heard so far from them was always that a mod can probably avoid to send it.

As there exists a solution and it has no downsides AFAIK, why not just adopt that?

**Comment by appgurueu** [comment]
> As there exists a solution and it has no downsides AFAIK, why not just adopt that?

It has a major code quality downside: The check is at the wrong place. Everywhere something modifies the HUD, it will have to compare whether something changed. This should be done by the engine to minimize code duplication (and to maximize speed, ideally). Temporarily, a wrapper that avoids resending unchanged HUD flags can be created; that shouldn't be the business of the minimap mod.

**Comment by erlehmann** [comment]
> The check is at the wrong place.

I disagree. The proposed check is at the place where it is decided to do something.

To not do something instead of doing it wrong and canceling it is a waste of CPU time.

> Everywhere something modifies the HUD, it will have to compare whether something changed.

Yes, this is indeed true for *everything* that wants to modify the HUD often and repeatedly.

So what?

> This should be done by the engine to minimize code duplication (and to maximize speed, ideally).

I agree that this should ideally be done by the engine – but mostly as a band-aid for shitty code.

> Temporarily, a wrapper that avoids resending unchanged HUD flags can be created; that shouldn't be the business of the minimap mod.

Such a wrapper would work, but would also be mostly a band-aid: The fastest code is the one never executed.

**Comment by appgurueu** [comment]
Side note, I believe this triggers a `hud_changed` event in builtin which then leads to `update_builtin_statbars` which will send hud change packets for health and breath bar.

**Comment by lhofhansl** [comment]
A looong time ago I fixed a bug in the nssm mod (explanation here, got merged in tenPlus1's fork later: https://github.com/NPXcoot/nssm/issues/22). 

Not the same, but similar problem, where a mod can send block updates that starve out all other block updates from the server (the client stopped loading any other blocks). Definitely something the engine should fix - to some extend (f.e. not sending blocks that have not changed).
